### PR TITLE
Add support for Windows

### PR DESCRIPTION
**Pull request by bioball**
Closes https://github.com/apple/pkl/issues/20

This adds support for Windows.
The in-language path separator is still `/`, to ensure Pkl programs are cross-platform.
The specifics behind this design are detailed in https://github.com/apple/pkl-evolution/pull/7.

Log lines are written using CRLF endings on Windows.
Modules that are combined with `--module-output-separator` uses LF endings to ensure
consistent rendering across platforms.

`jpkl` does not work on Windows as a direct executable.
However, it can work with `java -jar jpkl`.

Additional details:

* Adjust git settings for Windows
* Add native executable for pkl cli
* Add jdk17 windows Gradle check in CI
* Adjust CI test reports to be staged within Gradle rather than by shell script.
* Fix: encode more characters that are not safe Windows paths
* Skip running tests involving symbolic links on Windows (these require administrator privileges to run).
* Introduce custom implementation of `IoUtils.relativize`, because the current implementation delegates to `Path`, which will fail if the URI contains invalid Windows characters.
* Allow Gradle to initialize ExecutableJar `Property` values
* Add a way to enable remote debugging with `./gradlew test` commands
* Standard library `pkl:platform`'s `current.operatingSystem.name` is `"Windows"` for all Windows OSes (match treatment of macOS/Linux).

### CODE REVIEW

File: .circleci/jobs/GradleJob.pkl [hunk 1]
```diff
@@ -37,15 +41,6 @@ job: Config.Job = new {
...
     ...module.steps
-    new Config.RunStep {
-      // find all test results and write them to the home dir
-      name = "Gather test results"
-      command = """
-        mkdir ~/test-results/
-        find . -type f -regex ".*/build/test-results/.*xml" -exec cp {} ~/test-results/ \\;
-        """
-      `when` = "always"
-    }
```
Comment (bioball) [inline .circleci/jobs/GradleJob.pkl:48]: This step breaks on Windows.

Instead of trying to figure out how to write a Windows version, I added some logic to specify the test reports dir (see line 29).

File: .circleci/jobs/GradleJob.pkl [hunk 2]
```diff
@@ -15,14 +15,18 @@
...
 abstract module GradleJob
 
-import "package://pkg.pkl-lang.org/pkl-pantry/com.circleci.v2@1.1.0#/Config.pkl"
+import "package://pkg.pkl-lang.org/pkl-pantry/com.circleci.v2@1.1.2#/Config.pkl"
 
 /// Whether this is a release build or not.
 isRelease: Boolean = false
 
+/// The OS to run on
+os: "macOS"|"linux"|"windows"
```
Comment (bioball) [inline .circleci/jobs/GradleJob.pkl:24]: Moved this from `BuildNativeJob`, because this adds a `./gradlew check` job for Windows to PRBs.

File: buildSrc/src/main/kotlin/BuildInfo.kt [hunk 3]
```diff
@@ -85,9 +87,14 @@ open class BuildInfo(project: Project) {
...
-      Runtime.getRuntime()
-        .exec(arrayOf("git", "rev-parse", "--short", "HEAD"), arrayOf(), project.rootDir)
-        .inputStream.reader().readText().trim()
+      val process = ProcessBuilder()
+        .command("git", "rev-parse", "--short", "HEAD")
+        .directory(project.rootDir)
+        .start()
+      process.waitFor().also { exitCode ->
+        if (exitCode == -1) throw RuntimeException(process.errorStream.reader().readText())
+      }
```
Comment (bioball) [inline buildSrc/src/main/kotlin/BuildInfo.kt:96]: Tangential improvement: changed this because the failure was getting swallowed and I wasn't sure why the commit ID was not being populated correctly on Windows.

File: buildSrc/src/main/kotlin/ExecutableJar.kt [hunk 4]
```diff
@@ -15,27 +15,25 @@ import org.gradle.kotlin.dsl.listProperty
...
-  val inJar: RegularFileProperty = project.objects.fileProperty()
+  abstract val inJar: RegularFileProperty
 
   @get:OutputFile
-  val outJar: RegularFileProperty = project.objects.fileProperty()
+  abstract val outJar: RegularFileProperty
 
   @get:Input
-  val jvmArgs: ListProperty<String> = project.objects.listProperty()
+  abstract val jvmArgs: ListProperty<String>
```
Comment (bioball) [inline buildSrc/src/main/kotlin/ExecutableJar.kt:26]: This change is unrelated to Windows. Happy to move this to a separate PR if you'd prefer.

File: pkl-cli/src/main/kotlin/org/pkl/cli/CliEvaluator.kt [hunk 5]
```diff
@@ -158,7 +160,7 @@ constructor(
           } else {
             if (output.isNotEmpty()) {
               outputFile.writeString(
-                options.moduleOutputSeparator + IoUtils.getLineSeparator(),
+                options.moduleOutputSeparator + '\n',
```
Comment (bioball) [inline pkl-cli/src/main/kotlin/org/pkl/cli/CliEvaluator.kt:163]: Changed this to be `LF` and not platform dependent. Otherwise, `pkl eval foo.pkl bar.pkl` will give you different results on different OSes, which seems quite undesirable.

File: pkl-cli/src/test/kotlin/org/pkl/cli/CliProjectPackagerTest.kt [hunk 6]
```diff
@@ -858,17 +867,18 @@ class CliProjectPackagerTest {
...
-      .out/project2@2.0.0/project2@2.0.0
-      .out/project2@2.0.0/project2@2.0.0.sha256
+      .out${sep}project1@1.0.0${sep}project1@1.0.0.zip
+      .out${sep}project1@1.0.0${sep}project1@1.0.0.zip.sha256
+      .out${sep}project1@1.0.0${sep}project1@1.0.0
+      .out${sep}project1@1.0.0${sep}project1@1.0.0.sha256
+      .out${sep}project2@2.0.0${sep}project2@2.0.0.zip
+      .out${sep}project2@2.0.0${sep}project2@2.0.0.zip.sha256
+      .out${sep}project2@2.0.0${sep}project2@2.0.0
+      .out${sep}project2@2.0.0${sep}project2@2.0.0.sha256
```
Comment (bioball) [inline pkl-cli/src/test/kotlin/org/pkl/cli/CliProjectPackagerTest.kt:881]: Log lines on Windows will output CRLF endings, and also output `\` as the path separator.

The rationale is that these are log lines, which are meant to be useful for the host machine specifically.

File: pkl-core/src/main/java/org/pkl/core/module/ModuleKeyFactories.java [hunk 7]
```diff
@@ -141,15 +141,20 @@ public Optional<ModuleKey> create(URI uri) {
   private static class File implements ModuleKeyFactory {
     @Override
     public Optional<ModuleKey> create(URI uri) {
-      Path path;
-      try {
-        path = Path.of(uri);
```
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/module/ModuleKeyFactories.java:146]: Needed to adjust this because `Path.of("foo:bar")` throws an IllegalArgumentException, but yet should be treated as a file module key.

Changed this to initialize `Path` within the module key logic rather than here.

File: .gitattributes [hunk 8]
```diff
@@ -4,3 +4,4 @@
 /docs/** linguist-documentation
 
 *.pkl linguist-language=Groovy
+* text eol=lf
```
Comment (bioball) [inline .gitattributes:7]: Needed for git, and for spotless apply.

**Review by stackoverflow — APPROVED** [review]
Looks good.

File: docs/modules/pkl-cli/pages/index.adoc [hunk 9]
```diff
@@ -175,45 +178,44 @@ Pkl {pkl-version} (Linux, native)
...
+=== Windows Executable
 
 [source,shell]
 [subs="+attributes"]
 ----
-curl -L -o jpkl '{uri-pkl-java-download}'
-chmod +x jpkl
-./jpkl --version
+Invoke-WebRequest '{uri-pkl-windows-download}' -OutFile pkl.exe
+chmod +x pkl.exe
```
[the remark this answers is not in the document]
Comment (bioball) [reply docs/modules/pkl-cli/pages/index.adoc:187]: Oh, good catch.

File: pkl-core/src/main/java/org/pkl/core/project/ProjectPackager.java [hunk 10]
```diff
@@ -363,7 +370,7 @@ private List<Path> collectPackageElements(Project project, Package pkg) {
   }
 
   private boolean isAbsoluteImport(String importStr) {
-    return importStr.matches("\\w:.*") || importStr.startsWith("@");
+    return importStr.matches("\\w+:.*") || importStr.startsWith("@");
```
[the remark this answers is not in the document]
Comment (bioball) [reply pkl-core/src/main/java/org/pkl/core/project/ProjectPackager.java:373]: Yup

File: pkl-core/src/main/java/org/pkl/core/util/IoUtils.java [hunk 11]
```diff
@@ -180,6 +192,10 @@ public static String getLineSeparator() {
     return System.getProperty("line.separator");
   }
 
+  public static Boolean isWindows() {
+    return System.getProperty("os.name").contains("Windows");
```
[the remark this answers is not in the document]
Comment (bioball) [reply pkl-core/src/main/java/org/pkl/core/util/IoUtils.java]: I'll switch this to use `Platform.current().operatingSystem()`, which avoids the `contains` call every time.

File: pkl-core/src/main/java/org/pkl/core/module/ResolvedModuleKeys.java [hunk 12]
```diff
@@ -75,7 +77,16 @@ public URI getUri() {
...
     @Override
     public String loadSource() throws IOException {
-      return Files.readString(path, StandardCharsets.UTF_8);
+      try {
+        return Files.readString(path, StandardCharsets.UTF_8);
+      } catch (AccessDeniedException e) {
+        // Windows throws `AccessDeniedException` when reading directories.
+        // Sync error between different OSes.
+        if (Files.isDirectory(path)) {
+          throw new IOException("Is a directory");
```
Comment (holzensp) [inline pkl-core/src/main/java/org/pkl/core/module/ResolvedModuleKeys.java:86]: Should we not preserve the trace? I'd, at least, add `e` as a cause.
```suggestion
          throw new IOException("Is a directory", e);
```

File: pkl-doc/src/test/kotlin/org/pkl/doc/CliDocGeneratorTest.kt [hunk 13]
```diff
@@ -219,6 +226,11 @@ class CliDocGeneratorTest {
       .withFailMessage("Test bug: $actualFile should exist but does not.")
       .exists()
 
+    // symlinks on Git and Windows is rather finnicky; they create shortcuts by default unless
+    // a core Git option is set. Also, by default, symlinks require administrator privileges to run.
```
Comment (holzensp) [inline pkl-doc/src/test/kotlin/org/pkl/doc/CliDocGeneratorTest.kt:230]: ... but a `.lnk` is a text file with the redirected path in, no? Do you also need admin to read? If not; Is it an option to read `actualFile` instead and replace it with the contents? (Would possibly require a `while`, because of link-to-link cases.)

File: pkl-executor/src/main/java/org/pkl/executor/EmbeddedExecutor.java [hunk 14]
```diff
@@ -157,7 +157,15 @@ private PklDistribution findCompatibleDistribution(
 
   private static Path toDisplayPath(Path modulePath, ExecutorOptions options) {
     var rootDir = options.getRootDir();
-    return rootDir == null ? modulePath : rootDir.relativize(modulePath);
+    return rootDir == null ? modulePath : relativize(modulePath, rootDir);
+  }
+
+  // Windows relativize will fail if the two paths have different roots.
+  private static Path relativize(Path path, Path base) {
+    if (path.isAbsolute() && base.isAbsolute() && !path.getRoot().equals(base.getRoot())) {
```
Comment (holzensp) [inline pkl-executor/src/main/java/org/pkl/executor/EmbeddedExecutor.java:165]: I'd be inclined to push it in
```suggestion
    return relativize(modulePath, rootDir);
  }

  // Windows relativize will fail if the two paths have different roots.
  private static Path relativize(Path path, Path base) {
    if (base == null || path.isAbsolute() && base.isAbsolute() && !path.getRoot().equals(base.getRoot())) {
```

File: .circleci/config.pkl [hunk 15]
```diff
@@ -134,6 +136,11 @@ local buildNativeJobs: Mapping<String, BuildNativeJob> = new {
       musl = true
       isRelease = _dist == "release"
     }
+    ["pkl-cli-windows-amd64-\(_dist)"] {
```
[the remark this answers is not in the document]
Comment (bioball) [reply .circleci/config.pkl:139]: GraalVM doesn't have Windows ARM support. I'm actually not sure if they plan on supporting it; my question about it in Slack has not received any responses: https://graalvm.slack.com/archives/CNBFR78F9/p1715635155814669

File: pkl-cli/src/test/kotlin/org/pkl/cli/CliEvaluatorTest.kt [hunk 16]
```diff
@@ -978,6 +984,56 @@ result = someLib.x
...
+    val moduleUri =
+      writePklFile(
+        "test.pkl",
+        """
+      output {
+        files {
+          ["foo:bar"] { text = "bar" }
+        }
+      }
+    """
```
[the remark this answers is not in the document]
Comment (bioball) [reply pkl-cli/src/test/kotlin/org/pkl/cli/CliEvaluatorTest.kt:999]: Yes

File: pkl-core/src/main/java/org/pkl/core/module/ModuleKeyFactories.java (hunk 7 above)
Comment (bioball) [reply pkl-core/src/main/java/org/pkl/core/module/ModuleKeyFactories.java:146]: Yup; this is what we get on Windows:

```
jshell> Path.of(URI.create("file:///foo:bar"))
|  Exception java.nio.file.InvalidPathException: Illegal char <:> at index 4: /foo:bar
|        at WindowsPathParser.normalize (WindowsPathParser.java:182)
|        at WindowsPathParser.parse (WindowsPathParser.java:153)
|        at WindowsPathParser.parse (WindowsPathParser.java:77)
|        at WindowsPath.parse (WindowsPath.java:92)
|        at WindowsUriSupport.fromUri (WindowsUriSupport.java:166)
|        at WindowsFileSystemProvider.getPath (WindowsFileSystemProvider.java:98)
|        at Path.of (Path.java:203)
|        at (#6:1)
```

File: docs/modules/pkl-cli/pages/index.adoc [hunk 17]
```diff
@@ -175,45 +178,44 @@ Pkl {pkl-version} (Linux, native)
...
+.\pkl.exe --version
```
[the remark this answers is not in the document]
Comment (bioball) [reply docs/modules/pkl-cli/pages/index.adoc:187]: `.\` is needed. Looks like `.exe` can be omitted, though.

File: pkl-doc/src/test/kotlin/org/pkl/doc/CliDocGeneratorTest.kt (hunk 13 above)
Comment (bioball) [reply pkl-doc/src/test/kotlin/org/pkl/doc/CliDocGeneratorTest.kt:230]: As far as I can tell, there's no easy way to parse a `lnk` file in Java and figure out where it's pointed to.

File: pkl-cli/src/main/kotlin/org/pkl/cli/CliEvaluator.kt (hunk 5 above)
Comment (bioball) [reply pkl-cli/src/main/kotlin/org/pkl/cli/CliEvaluator.kt:163]: This is configuration output meant to be saved to a file, or applied to a system.

The standard use-case here is to generate a YAML document stream. For example, `pkl eval podA.pkl podB.pkl > output.yaml`.

Note that if the `-o` option is set, it already uses `LF` endings. So, without this change, these two would have had different results:

```
pkl eval podA.pkl podB.pkl > output.yaml
```

```
pkl eval podA.pkl podB.pkl -o output.yaml
```

File: pkl-cli/src/test/kotlin/org/pkl/cli/CliProjectPackagerTest.kt (hunk 6 above)
Comment (bioball) [reply pkl-cli/src/test/kotlin/org/pkl/cli/CliProjectPackagerTest.kt:881]: Can you elaborate on what is not consistent here?

File: pkl-core/src/main/java/org/pkl/core/module/ResolvedModuleKeys.java (hunk 12 above)
Comment (bioball) [reply pkl-core/src/main/java/org/pkl/core/module/ResolvedModuleKeys.java:86]: While this is generally a good idea, it works against us here because the cause gets rendered in the `-- Pkl Error --` message. The idea here is that we want to make this error message identical on the two platforms.

File: pkl-core/src/main/java/org/pkl/core/module/ModuleKeys.java [hunk 18]
```diff
@@ -316,7 +316,13 @@ public List<PathElement> listElements(SecurityManager securityManager, URI baseU
     public ResolvedModuleKey resolve(SecurityManager securityManager)
         throws IOException, SecurityManagerException {
       securityManager.checkResolveModule(uri);
-      var realPath = path.toRealPath();
+      // Disallow paths that contain `\\` characters if on Windows
+      // (require `/` as the path separator on all OSes)
+      var uriPath = uri.getPath();
+      if (java.io.File.separatorChar == '\\' && uriPath != null && uriPath.contains("\\")) {
+        throw new FileNotFoundException();
```
[the remark this answers is not in the document]
Comment (bioball) [reply pkl-core/src/main/java/org/pkl/core/module/ModuleKeys.java:323]: Fair point; will add an extra hint here about it. But I'm thinking let's just show this on every OS. It may or may not be useful to users on *nix platforms that are used to Windows style paths. Also, having this be Windows only greatly complicates our testing.

File: pkl-executor/src/main/java/org/pkl/executor/EmbeddedExecutor.java (hunk 14 above)
Comment (bioball) [reply pkl-executor/src/main/java/org/pkl/executor/EmbeddedExecutor.java:165]: I actually find the former a little easier to read, because of fewer binary operators grouped in the `if`
